### PR TITLE
Homework 1: Henning Richter

### PR DESCRIPTION
**Pull request by hediriro**
Ok, submission... maybe this works here

### CODE REVIEW

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 1]
```diff
@@ -1,13 +1,52 @@
...
+        int imageSizeRow = image.length;
+        int imageSizeColumn = image[0].length;
+        Color[][] output = new Color[imageSizeRow][imageSizeColumn];
+
+        for (int imageRow = 0; imageRow < imageSizeRow; imageRow++) {
+            for (int imageColumn = 0; imageColumn < imageSizeColumn; imageColumn++) {
+
+                int redNew = 0;
+                int greenNew = 0;
+                int blueNew=0;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: Don't forget about code style ;)

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 2]
```diff
@@ -1,13 +1,52 @@
...
+
+                for (int kernelRow = 0; kernelRow < kernelSizeRow; kernelRow++) {
+                    for (int kernelColumn = 0; kernelColumn < kernelSizeColumn; kernelColumn++) {
+
+                        int rowNavigator = kernelRow - kernelRadiusRow; // can be -1, 0, +1 for 3x3 kernel
+                        int columnNavigator = kernelColumn - kernelRadiusColumn; // can be -1, 0, +1 for 3x3 kernel
+
+                        // if core in image range ... else do nothing, as * with 0 is always 0
+                        if ((imageRow + rowNavigator > -1) && (imageRow + rowNavigator < output.length) && (imageColumn + columnNavigator > -1) && (imageColumn + columnNavigator < output[imageRow].length)) {
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: Extract variable `imageRow + rowNavigator`

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 3]
```diff
@@ -1,13 +1,52 @@
...
+                            int redTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getRed();
+                            int greenTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getGreen();
+                            int blueTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getBlue();
+                            redTemp = (int)Math.floor(redTemp * kernel[kernelRow][kernelColumn]);
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: You can just cast it to int without Math.floor, it is useless operation

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 4]
```diff
@@ -1,13 +1,52 @@
...
+                            greenTemp = (int)Math.floor(greenTemp * kernel[kernelRow][kernelColumn]);
+                            blueTemp = (int)Math.floor(blueTemp * kernel[kernelRow][kernelColumn]);
+                            redNew = redNew + redTemp;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: You can remove redTemp and use `+=` operator

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 5]
```diff
@@ -1,13 +1,54 @@
...
+                        int imageRowCorrection = imageRow + rowNavigator;
+                        int imageColumnCorrection = imageColumn + columnNavigator;
+                        if ((imageRowCorrection > -1) && (imageRowCorrection < output.length) && (imageColumnCorrection > -1) && (imageColumnCorrection < output[imageRow].length)) {
+                            int redTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getRed();
+                            int greenTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getGreen();
+                            int blueTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getBlue();
+                            redTemp = (int)(redTemp * kernel[kernelRow][kernelColumn]);
+                            greenTemp = (int)(greenTemp * kernel[kernelRow][kernelColumn]);
+                            blueTemp = (int)(blueTemp * kernel[kernelRow][kernelColumn]);
+                            redNew += redTemp;
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: I mean that you can remove `redTemp` and use `renNew += (int) ...`

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 6]
```diff
@@ -1,13 +1,54 @@
...
+
+                        int rowNavigator = kernelRow - kernelRadiusRow; // can be -1, 0, +1 for 3x3 kernel
+                        int columnNavigator = kernelColumn - kernelRadiusColumn; // can be -1, 0, +1 for 3x3 kernel
+
+                        // if core in image range ... else do nothing, as * with 0 is always 0
+                        int imageRowCorrection = imageRow + rowNavigator;
+                        int imageColumnCorrection = imageColumn + columnNavigator;
+                        if ((imageRowCorrection > -1) && (imageRowCorrection < output.length) && (imageColumnCorrection > -1) && (imageColumnCorrection < output[imageRow].length)) {
+                            int redTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getRed();
+                            int greenTemp = image[imageRow + rowNavigator][imageColumn + columnNavigator].getGreen();
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: You have already extracted `imageRowCorrection` and `imageColumnCorrection` but don't use it here

Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: You can extract variable with `image[imageRow + rowNavigator][imageColumn + columnNavigator]`

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 7]
```diff
@@ -1,13 +1,57 @@
...
+
+                        int rowNavigator = kernelRow - kernelRadiusRow; // can be -1, 0, +1 for 3x3 kernel
+                        int columnNavigator = kernelColumn - kernelRadiusColumn; // can be -1, 0, +1 for 3x3 kernel
+
+                        // if core in image range ... else do nothing, as * with 0 is always 0
+                        int imageRowCorrection = imageRow + rowNavigator;
+                        int imageColumnCorrection = imageColumn + columnNavigator;
+                        if (isInRange(imageRowCorrection, imageSizeRow, imageColumnCorrection, imageSizeColumn)) {
+                            Color pixel = image[imageRowCorrection][imageColumnCorrection];
+                            int redTemp = pixel.getRed();
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: These temp variables are redundant

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 8]
```diff
@@ -1,13 +1,57 @@
...
+                            int greenTemp = pixel.getGreen();
+                            int blueTemp = pixel.getBlue();
+                            redNew += (int)(redTemp * kernel[kernelRow][kernelColumn]);
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: `kernel[kernelRow][kernelColumn]` can be extracted

File: src/main/java/impl/image/ConvolutionProviderImpl.java [hunk 9]
```diff
@@ -1,13 +1,57 @@
...
+                        }
+                    }
+                }
+                output[imageRow][imageColumn] = new Color(redNew, greenNew, blueNew, 255);
+            }
+        }
+        return output;
+    }
+
+    private boolean isInRange(int imageRowCorrection, int imageSizeRow, int imageColumnCorrection, int imageSizeColumn){
```
Comment (Walingar) [inline src/main/java/impl/image/ConvolutionProviderImpl.java]: code-style

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/image/ConvolutionProviderImpl.java (hunk 8 above)
Comment (hediriro) [reply src/main/java/impl/image/ConvolutionProviderImpl.java]: I forgot to apply this before re-requesting review, sorry! However, I already applied it afterwards.

**Review by Walingar — APPROVED** [review]

**Pull request merged**